specs/keccak256/variable-length-input.md

```diff
@@ -0,0 +1,147 @@
+# Keccak256 For Variable Length Input
+
+We describe the circuit that can process inputs of an arbitrary size.
+
+## Plain Keccak behavior
+
+Keccak256 uses a simple iterating sponge construction to handle a variable length input. The sponge absorbs 17 words(136 bytes) of inputs, applies the `Keccak-f` permutation, and then absorb next 17 words. Once all parts of the inputs are absorbed, the sponge squeezes out the 32 bytes for the output.
+
+If the last part of the input is less than 136 bytes, a padding must be applied. The padding is a multi-rate padding, it pads single bit 1 followed by minimum number of bits 0 followed by a single bit 1, such that the result is 136 bytes.
+
+## Circuit behavior
+
+### Lookup Region
+
+This region splits the input to multiple parts, each part corresponds to their `Keccak-f` permutation round.
+This is also a lookup table for the other circuits to lookup the Keccak256 input to the output.
+
+Define:
+
+We use `Round` for the cells of the following columns in a row, it represents the information in a `keccak_f` round.
+
+The `curr: Round` and `next: Round` represent the current and the next row.
+
+Columns:
+
+- `hash_id` Sequential number for grouping different hash
+- `input_len` Length for correct padding
+- `input` 136 bytes to be absorbed in this round
+- `acc_len` How many length we have absorbed
+- `acc_input` Accumulatd bytes by random linear combination (in big-endian order)
```

Comment:
In this context, aren't these the same? We have absorbed as many bytes as we have accomulated in the RLC no?

Reply:
The `acc_len` is the length of bytes, while `acc_input` is RLC. `acc_len` can be 0 or a multiple of 136, we use it to track if a hash input ends at `input_len`.

Reply:
ˋacc_inputˋ is always [0,135]??